File: AkaiS950List/Hfe.cs
using System;
using System.Collections.Generic;

namespace AkaiS950List
{
    /// <summary>A sector recovered from the MFM bitstream.</summary>
    public sealed class Sect
    {
        public int Cyl, Head, Sec, SizeCode;
        public byte[] Data;
        public bool IdCrcOk, DataCrcOk;
    }

    /// <summary>
    /// Decoder for HxC/Gotek .hfe images (HXCPICFE rev 0). The image stores raw MFM
    /// cell data for both sides interleaved in 256-byte chunks; this reads the cells,
    /// finds the A1A1A1 sync marks, decodes the ID and data fields and verifies CRCs.
    /// </summary>
    public static class Hfe
    {
        static ushort U16(byte[] b, int o) { return (ushort)(b[o] | (b[o + 1] << 8)); }

        /// <summary>Pull one side's cell bytes out of the interleaved track blocks.</summary>
        public static byte[] SideCells(byte[] img, int track, int side)
        {
            int lut = U16(img, 18) * 512;
            if (lut + track * 4 + 3 >= img.Length) return new byte[0];
            int tOff = U16(img, lut + track * 4) * 512;
            int tLen = U16(img, lut + track * 4 + 2);
            int half = tLen / 2;
            if (half <= 0) return new byte[0];

            var o = new byte[half];
            int w = 0, pos = tOff;
            while (w < half)
            {
                int chunk = Math.Min(256, half - w);
                int src = pos + (side == 0 ? 0 : 256);
                if (src < 0 || src + chunk > img.Length) break;
                Array.Copy(img, src, o, w, chunk);
                w += chunk;
                pos += 512;
            }
            return o;
        }

        /// <summary>One byte per MFM cell; HFE stores the first cell in bit 0.</summary>
        static byte[] Unpack(byte[] cells)
        {
            var bits = new byte[cells.Length * 8];
            for (int i = 0; i < cells.Length; i++)
            {
                byte c = cells[i];
                for (int b = 0; b < 8; b++) bits[i * 8 + b] = (byte)((c >> b) & 1);
            }
            return bits;
        }

        // In MFM the cells alternate clock, data; the data bits sit at odd offsets.
        static byte ReadByte(byte[] bits, int groupStart)
        {
            int n = bits.Length, v = 0;
            for (int i = 0; i < 8; i++) v = (v << 1) | bits[(groupStart + 2 * i + 1) % n];
            return (byte)v;
        }

        static ushort Crc(byte[] d, int len, ushort crc)
        {
            for (int i = 0; i < len; i++)
            {
                crc ^= (ushort)(d[i] << 8);
                for (int b = 0; b < 8; b++)
                    crc = (ushort)(((crc & 0x8000) != 0) ? ((crc << 1) ^ 0x1021) : (crc << 1));
            }
            return crc;
        }

        static readonly byte[] A1x3 = new byte[] { 0xA1, 0xA1, 0xA1 };

        public static List<Sect> DecodeTrack(byte[] cells)
        {
            var outp = new List<Sect>();
            if (cells.Length == 0) return outp;

            byte[] bits = Unpack(cells);
            int n = bits.Length;
            uint sr = 0;
            int lastSync = -100, run = 0;
            Sect pendingId = null;

            for (int p = 0; p < n; p++)
            {
                sr = (sr << 1) | bits[p];
                if ((sr & 0xFFFF) != 0x4489) continue;   // A1 with a missing clock

                run = (p - lastSync == 16) ? run + 1 : 1;
                lastSync = p;
                if (run < 3) continue;

                int g = p + 1;
                byte mark = ReadByte(bits, g); g += 16;

                if (mark == 0xFE)                        // ID address mark
                {
                    var id = new byte[4];
                    for (int i = 0; i < 4; i++) { id[i] = ReadByte(bits, g); g += 16; }
                    int cs = ReadByte(bits, g) << 8; g += 16;
                    cs |= ReadByte(bits, g); g += 16;

                    ushort c = Crc(A1x3, 3, 0xFFFF);
                    c = Crc(new[] { mark }, 1, c);
                    c = Crc(id, 4, c);

                    pendingId = new Sect
                    {
                        Cyl = id[0], Head = id[1], Sec = id[2], SizeCode = id[3],
                        IdCrcOk = (c == cs)
                    };
                }
                else if ((mark == 0xFB || mark == 0xF8) && pendingId != null)  // data mark
                {
                    int sz = 128 << (pendingId.SizeCode & 7);
                    var data = new byte[sz];
                    for (int i = 0; i < sz; i++) { data[i] = ReadByte(bits, g); g += 16; }
                    int cs = ReadByte(bits, g) << 8; g += 16;
                    cs |= ReadByte(bits, g); g += 16;

                    ushort c = Crc(A1x3, 3, 0xFFFF);
                    c = Crc(new[] { mark }, 1, c);
                    c = Crc(data, sz, c);

                    pendingId.Data = data;
                    pendingId.DataCrcOk = (c == cs);
                    outp.Add(pendingId);
                    pendingId = null;
                }
            }
            return outp;
        }

        /// <summary>
        /// Decode every track and lay the sectors out linearly:
        /// LBA = (cyl * sides + head) * 5 + (sec - 1), 1024 bytes each.
        /// </summary>
        public static byte[] Extract(byte[] raw, out int badCrc, out int missing)
        {
            const int Spt = 5, Ssz = 1024;
            int tracks = raw[9], sides = raw[10];
            var img = new byte[tracks * sides * Spt * Ssz];
            var got = new bool[tracks * sides * Spt];
            badCrc = 0;

            for (int t = 0; t < tracks; t++)
                for (int s = 0; s < sides; s++)
                    foreach (var sec in DecodeTrack(SideCells(raw, t, s)))
                    {
                        if (sec.Data == null) continue;
                        if (sec.Sec < 1 || sec.Sec > Spt || sec.Cyl >= tracks || sec.Head >= sides) continue;

                        int lba = (sec.Cyl * sides + sec.Head) * Spt + (sec.Sec - 1);
                        if (lba < 0 || lba >= got.Length || got[lba]) continue;   // keep the first good copy

                        if (!sec.DataCrcOk || !sec.IdCrcOk) badCrc++;
                        Array.Copy(sec.Data, 0, img, lba * Ssz, Math.Min(Ssz, sec.Data.Length));
                        got[lba] = true;
                    }

            missing = 0;
            foreach (bool g in got) if (!g) missing++;
            return img;
        }
    }
}

File: Plugin/Source/S950/Hfe.cpp
#include "Hfe.h"

#include <algorithm>
#include <cstring>

namespace s950::hfe
{
    namespace
    {
        std::uint16_t u16 (const std::vector<unsigned char>& b, std::size_t at)
        {
            if (at + 1 >= b.size()) return 0;
            return static_cast<std::uint16_t> (b[at] | (b[at + 1] << 8));
        }

        /// One byte per MFM cell; HFE stores the first cell in bit 0.
        std::vector<unsigned char> unpack (const std::vector<unsigned char>& cells)
        {
            std::vector<unsigned char> bits (cells.size() * 8);

            for (std::size_t i = 0; i < cells.size(); ++i)
            {
                const unsigned char c = cells[i];
                for (int b = 0; b < 8; ++b)
                    bits[i * 8 + static_cast<std::size_t> (b)] = static_cast<unsigned char> ((c >> b) & 1);
            }

            return bits;
        }

        /*
         * In MFM the cells alternate clock, data, so the data bits sit at odd offsets.
         *
         * The modulo is not caution: a sector's last field can run off the end of the track
         * and wrap round to its start, which is what the physical disk does too.
         */
        unsigned char readByte (const std::vector<unsigned char>& bits, std::size_t groupStart)
        {
            const std::size_t n = bits.size();
            if (n == 0) return 0;

            int v = 0;
            for (std::size_t i = 0; i < 8; ++i)
                v = (v << 1) | bits[(groupStart + 2 * i + 1) % n];

            return static_cast<unsigned char> (v);
        }

        /// CRC-16/CCITT, as the floppy format uses it.
        std::uint16_t crc16 (const unsigned char* d, std::size_t len, std::uint16_t crc)
        {
            for (std::size_t i = 0; i < len; ++i)
            {
                crc ^= static_cast<std::uint16_t> (d[i] << 8);

                for (int b = 0; b < 8; ++b)
                    crc = static_cast<std::uint16_t> ((crc & 0x8000) != 0 ? ((crc << 1) ^ 0x1021)
                                                                         : (crc << 1));
            }

            return crc;
        }

        struct Sector
        {
            int  cyl = 0, head = 0, sec = 0, sizeCode = 0;
            std::vector<unsigned char> data;
            bool idCrcOk = false, dataCrcOk = false;
        };

        const unsigned char A1x3[3] = { 0xA1, 0xA1, 0xA1 };

        std::vector<Sector> decodeTrack (const std::vector<unsigned char>& cells)
        {
            std::vector<Sector> out;
            if (cells.empty()) return out;

            const auto bits = unpack (cells);
            const std::size_t n = bits.size();

            std::uint32_t sr = 0;
            long long lastSync = -100;
            int  run = 0;

            bool   havePending = false;
            Sector pending;

            for (std::size_t p = 0; p < n; ++p)
            {
                sr = (sr << 1) | bits[p];

                // A1 written with a missing clock bit - the mark that says a field follows.
                if ((sr & 0xFFFF) != 0x4489) continue;

                run = (static_cast<long long> (p) - lastSync == 16) ? run + 1 : 1;
                lastSync = static_cast<long long> (p);
                if (run < 3) continue;

                std::size_t g = p + 1;
                const unsigned char mark = readByte (bits, g);
                g += 16;

                if (mark == 0xFE)                                   // ID address mark
                {
                    unsigned char id[4];
                    for (int i = 0; i < 4; ++i) { id[i] = readByte (bits, g); g += 16; }

                    int cs = readByte (bits, g) << 8; g += 16;
                    cs |= readByte (bits, g);         g += 16;

                    std::uint16_t c = crc16 (A1x3, 3, 0xFFFF);
                    c = crc16 (&mark, 1, c);
                    c = crc16 (id, 4, c);

                    pending = Sector();
                    pending.cyl      = id[0];
                    pending.head     = id[1];
                    pending.sec      = id[2];
                    pending.sizeCode = id[3];
                    pending.idCrcOk  = (c == cs);
                    havePending = true;
                }
                else if ((mark == 0xFB || mark == 0xF8) && havePending)   // data mark
                {
                    const std::size_t sz = static_cast<std::size_t> (128) << (pending.sizeCode & 7);

                    std::vector<unsigned char> data (sz);
                    for (std::size_t i = 0; i < sz; ++i) { data[i] = readByte (bits, g); g += 16; }

                    int cs = readByte (bits, g) << 8; g += 16;
                    cs |= readByte (bits, g);         g += 16;

                    std::uint16_t c = crc16 (A1x3, 3, 0xFFFF);
                    c = crc16 (&mark, 1, c);
                    c = crc16 (data.data(), sz, c);

                    pending.data      = std::move (data);
                    pending.dataCrcOk = (c == cs);

                    out.push_back (std::move (pending));
                    havePending = false;
                }
            }

            return out;
        }
    }

    /// One side's cell bytes, pulled out of the interleaved track blocks.
    std::vector<unsigned char> sideCells (const std::vector<unsigned char>& img,
                                          int track, int side)
    {
        const std::size_t lut = static_cast<std::size_t> (u16 (img, 18)) * 512;
        if (lut + static_cast<std::size_t> (track) * 4 + 3 >= img.size())
            return {};

        const std::size_t tOff = static_cast<std::size_t> (u16 (img, lut + static_cast<std::size_t> (track) * 4)) * 512;
        const int         tLen = u16 (img, lut + static_cast<std::size_t> (track) * 4 + 2);
        const int         half = tLen / 2;

        if (half <= 0) return {};

        std::vector<unsigned char> out (static_cast<std::size_t> (half));

        int         w   = 0;
        std::size_t pos = tOff;

        while (w < half)
        {
            const int         chunk = std::min (256, half - w);
            const std::size_t src   = pos + (side == 0 ? 0 : 256);

            if (src + static_cast<std::size_t> (chunk) > img.size()) break;

            std::memcpy (out.data() + w, img.data() + src, static_cast<std::size_t> (chunk));
            w   += chunk;
            pos += 512;
        }

        return out;
    }

    bool looksLikeHfe (const std::vector<unsigned char>& raw)
    {
        return raw.size() > 8 && std::equal (raw.begin(), raw.begin() + 8, "HXCPICFE");
    }

    std::vector<unsigned char> extract (const std::vector<unsigned char>& raw,
                                        int& badCrc, int& missing)
    {
        constexpr int Spt = 5, Ssz = 1024;

        badCrc  = 0;
        missing = 0;

        if (raw.size() < 12) return {};

        const int tracks = raw[9];
        const int sides  = raw[10];

        if (tracks <= 0 || sides <= 0) return {};

        const std::size_t sectors = static_cast<std::size_t> (tracks) * sides * Spt;

        std::vector<unsigned char> img (sectors * Ssz, 0);
        std::vector<bool>          got (sectors, false);

        for (int t = 0; t < tracks; ++t)
        {
            for (int s = 0; s < sides; ++s)
            {
                for (const auto& sec : decodeTrack (sideCells (raw, t, s)))
                {
                    if (sec.data.empty()) continue;
                    if (sec.sec < 1 || sec.sec > Spt) continue;
                    if (sec.cyl >= tracks || sec.head >= sides) continue;

                    const long long lba = (static_cast<long long> (sec.cyl) * sides + sec.head) * Spt
                                        + (sec.sec - 1);

                    if (lba < 0 || lba >= static_cast<long long> (sectors)) continue;
                    if (got[static_cast<std::size_t> (lba)]) continue;   // keep the first good copy

                    if (! sec.dataCrcOk || ! sec.idCrcOk) ++badCrc;

                    std::memcpy (img.data() + static_cast<std::size_t> (lba) * Ssz,
                                 sec.data.data(),
                                 std::min (static_cast<std::size_t> (Ssz), sec.data.size()));

                    got[static_cast<std::size_t> (lba)] = true;
                }
            }
        }

        for (bool g : got)
            if (! g) ++missing;

        return img;
    }
}
